### PR TITLE
feat(ecs): container dependencies

### LINKED ISSUE #2490: Unable to add `dependsOn` to container definitoin
Given there is no App Mesh support yet I'm trying to use L1 constructs. Although not critical I guess, the docs instruct adding a `dependsOn` Envoy proxy container to the app container definition.

Problem 1: there is no `cfnContainerDefinition`. Therefore I cannot use `addDependsOn()`

Problem 2: as an alternative trying to use `cfnTaskDefiniton_.AddOverride()` as below but it appears this function does handle arrays. `ContainerDefinitions` is an array, therefore I cannot see how to add a `dependsOn` node.

https://github.com/awslabs/aws-cdk/blob/66dd2282bb6879aff803f80531fb4a1d324e97e7/packages/%40aws-cdk/cdk/lib/cfn-resource.ts#L158

My C# code
```csharp  
public static void AddDependsOnProxy(TaskDefinition TaskDefinition, string AppName)
{
      var cfnTaskDef = (CfnTaskDefinition_)TaskDefinition.Node.FindChild("Resource"); 

      string jsonstring = @"[{'containerName': 'envoy', 'condition': 'HEALTHY'}]";

      cfnTaskDef.AddOverride("Properties.ContainerDefinitions[0].dependsOn", jsonstring);
}
```

### PR DESCRIPTION
**Pull request by Don-CA**
Add new `addContainerDependencies` method to allow for container dependencies
Fixes #2490

----
* [X] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
* [X] Docs
  - __jsdocs__: All public APIs documented
* [X] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

----

### CODE REVIEW

**Comment by cohalz** [comment]
:+1:

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 1]
```diff
@@ -562,6 +575,51 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
     hardLimit: ulimit.hardLimit,
   };
 }
+/**
+ * Container dependency
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Can you please add the link to ContainerDependency in the documentation: https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 2]
```diff
@@ -562,6 +575,51 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
...
+ */
+export interface ContainerDependency {
+  /**
+   * The container to depend on
+   */
+  readonly container: ContainerDefinition,
+  /**
+   * The dependency condition of the container
+   */
+  readonly condition?: ContainerDependencyCondition,
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: As per https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html condition is required, so you should either make condition required or provide a good default (preferable)

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 3]
```diff
@@ -562,6 +575,51 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
...
+   * This condition validates that the dependent container passes its Docker health check before permitting other containers to start.
+   * This requires that the dependent container has health checks configured. This condition is confirmed only at task startup.
+   */
+  HEALTHY = 'HEALTHY',
+}
+
+function renderContainerDependency(containerDependency: ContainerDependency): CfnTaskDefinition.ContainerDependencyProperty {
+  return {
+    containerName: containerDependency.container.node.id,
+    condition: containerDependency.condition || ContainerDependencyCondition.HEALTHY
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Looks like you're defining Healthy as the default, please document this as the default.

File: packages/@aws-cdk/aws-ecs/test/ec2/test.ec2-task-definition.ts [hunk 4]
```diff
@@ -128,7 +129,42 @@ export = {
 
       test.done();
     },
+    "correctly sets container dependenices"(test: Test) {
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs/test/ec2/test.ec2-task-definition.ts]: Can you also create a test where you specify a different container dependency condition (COMPLETE, SUCCESS)

File: packages/@aws-cdk/aws-apigateway/package-lock.json [hunk 5]
```diff
@@ -1,120 +0,0 @@
-{
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-apigateway/package-lock.json]: Why are you deleting this file?

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts (hunk 1 above)
Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Added that link and brought the documentation in line with the v1.0 style I saw elsewhere in the file.

File: packages/@aws-cdk/aws-apigateway/package-lock.json (hunk 5 above)
Comment (Don-CA) [reply packages/@aws-cdk/aws-apigateway/package-lock.json]: I have no idea! Restored.

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 6]
```diff
@@ -364,6 +369,13 @@ export class ContainerDefinition extends cdk.Construct {
     this.ulimits.push(...ulimits);
   }
 
+  /**
+   * This method adds one or more container dependencies to the container.
+   */
+  public addContainerDependencies(...containerDependencies: ContainerDependency[]) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Please also add an optional array property to the Props object to initialize these.

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 7]
```diff
@@ -418,6 +430,7 @@ export class ContainerDefinition extends cdk.Construct {
       command: this.props.command,
       cpu: this.props.cpu,
       disableNetworking: this.props.disableNetworking,
+      dependsOn: this.containerDependencies.map(renderContainerDependency),
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Please don't render the property at all if the list is empty.

Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Yeah, I was unsure about this. Rendering an empty list keeps in consistent with `Links`, `MountPoints` and `VolumesFrom` however it results in a big change. 
I'm happy to make it different to the other three. I'm happy to do a further PR to make the other three not output if their list is empty, would you be open to that?

Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Or, leave it as is, and refactor all four as a seperate PR?

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts (hunk 6 above)
Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: See my comment below, I'm keeping it consistent with similar properties but open to refactoring all the properties to output cleaner cloud formation.

File: packages/@aws-cdk/aws-codepipeline-actions/test/integ.pipeline-ecs-deploy.expected.json [hunk 8]
```diff
@@ -230,6 +230,7 @@
       "Properties": {
         "ContainerDefinitions": [
           {
+            "DependsOn": [],
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-actions/test/integ.pipeline-ecs-deploy.expected.json]: Preferably don't render if the array is empty

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 9]
```diff
@@ -238,6 +238,11 @@ export class ContainerDefinition extends cdk.Construct {
    */
   public readonly ulimits = new Array<Ulimit>();
 
+  /**
+   * An array dependencies defined for container startup and shutdown.
+   */
+  public readonly containerDependencies = Array<ContainerDependency>();
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: I am only now realizing that all of these are exposing mutable arrays to users... Not ideal. @rix0rrr what do you think?

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts (hunk 7 above)
Comment (eladb) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: We prefer not to render empty arrays/objects. Would be great to fix the others as well...

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 10]
```diff
@@ -589,6 +602,56 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
...
+/**
+ * The details of a dependency on another container in the task definition.
+ *
+ * @see https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html
+ */
+export interface ContainerDependency {
+  /**
+   * The container to depend on.
+   */
+  readonly container: ContainerDefinition,
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Use `;` for EOL

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 11]
```diff
@@ -589,6 +602,56 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
...
+  /**
+   * The dependency condition of the container. Valid values are ContainerDependencyCondition.START, ContainerDependencyCondition.COMPLETE,
+   * ContainerDependencyCondition.SUCCESS and ContainerDependencyCondition.HEALTHY.
+   *
+   * @default - ContainerDependencyCondition.HEALTHY
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: We only use `-` when it's not a "real" value. In your case you are actually providing a value...

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 12]
```diff
@@ -589,6 +602,56 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
...
+ *
+ * @see https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html
+ */
+export interface ContainerDependency {
+  /**
+   * The container to depend on.
+   */
+  readonly container: ContainerDefinition,
+  /**
+   * The dependency condition of the container. Valid values are ContainerDependencyCondition.START, ContainerDependencyCondition.COMPLETE,
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: What is the dependency condition? Can you copy & paste from AWS docs?

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts (hunk 7 above)
Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Ok, cool. I can fix the others as well as part of this PR.

**Comment by eladb** [comment]
See some review comments

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts (hunk 12 above)
Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: This is a copy & paste from the docs :) Added some more of my own context.

**Comment by Don-CA** [comment]
> Can you add tests to ensure that MountPoints, Links, etc still continue to render correctly if populated?
Added tests for VolumesFrom and Links.
There are already existing tests for the rest:

[MountPoints](https://github.com/aws/aws-cdk/blob/334261d417286d2e23c14561742b7cd3d76ecfff/packages/%40aws-cdk/aws-ecs/test/ec2/test.ec2-task-definition.ts#L163)

[Volumes] (https://github.com/aws/aws-cdk/blob/334261d417286d2e23c14561742b7cd3d76ecfff/packages/%40aws-cdk/aws-ecs/test/ec2/test.ec2-task-definition.ts#L171)

[PortMappings] (https://github.com/aws/aws-cdk/blob/334261d417286d2e23c14561742b7cd3d76ecfff/packages/%40aws-cdk/aws-ecs/test/ec2/test.ec2-task-definition.ts#L75)

[Ulimits] (https://github.com/aws/aws-cdk/blob/334261d417286d2e23c14561742b7cd3d76ecfff/packages/%40aws-cdk/aws-ecs/test/ec2/test.ec2-task-definition.ts#L80)

[

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 13]
```diff
@@ -469,15 +482,15 @@ export class ContainerDefinition extends cdk.Construct {
       image: this.imageConfig.imageName,
       memory: this.props.memoryLimitMiB,
       memoryReservation: this.props.memoryReservationMiB,
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Shouldn’t these be Lazy? What happens if you add something to the array after initialization?

File: packages/@aws-cdk/aws-ecs/lib/container-definition.ts [hunk 14]
```diff
@@ -640,6 +660,58 @@ function renderUlimit(ulimit: Ulimit): CfnTaskDefinition.UlimitProperty {
...
+  /**
+   * This condition validates that the dependent container passes its Docker health check before permitting other containers to start.
+   * This requires that the dependent container has health checks configured. This condition is confirmed only at task startup.
+   */
+  HEALTHY = 'HEALTHY',
+}
+
+function renderContainerDependency(containerDependency: ContainerDependency): CfnTaskDefinition.ContainerDependencyProperty {
+  return {
+    containerName: containerDependency.container.node.id,
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: If the name of the container is similar to the node ID that’s a detail that needs to be encapsulated behind a container.containerName property

Comment (Don-CA) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Would this also apply to the reference to `container.node.id` in `addLink`?

Comment (eladb) [reply packages/@aws-cdk/aws-ecs/lib/container-definition.ts]: Yes please!

**Comment by Don-CA** [comment]
> I really don’t like it that we expose all these public arrays. That’s an anti pattern because it exposes an internal mutable structure

I took all those arrays private and all the tests still ran, what do you think of doing that? It would create some breaking changes, would that be acceptable?

**Comment by eladb** [comment]
I don't think it will be acceptable to break this public API sadly. What we can do is mark them as @deprecated so people don't rely on them and we will get rid of it in v2.0 sometime in the future.

**Comment by eladb** [comment]
Apologies for the confusion... I was just venting ;-)

**Comment by Don-CA** [comment]
> Apologies for the confusion... I was just venting ;-)

No worries, reverted those changes :)

**Pull request merged**